lib/storage/data/external/AwsClient.js

```diff
@@ -140,78 +236,139 @@ class AwsClient {
-                err, this._dataStoreName, this.clientType);
-        });
-        // Always call the callback asynchronously: the caller may
-        // destroy the stream with destroy(), which MUST be
-        // asynchronous wrt. request.createReadStream() to avoid
-        // socket leaks, notably because of the request.abort() call.
-        return process.nextTick(() => callback(null, stream));
+        const command = new GetObjectCommand(params);
+        const abortController = new AbortController();
+        this._client.send(command, { abortSignal: abortController.signal })
+            .then(data => {
+                const rawBody = data.Body;
+                const stream = this._normalizeSdkStream(rawBody);
+                if (!stream || typeof stream.on !== 'function') {
+                    const bodyType = rawBody?.constructor?.name || typeof rawBody;
+                    log.error('unsupported body type from AWS SDK getObject response', {
+                        method: 'AwsClient.get',
+                        backendType: this.clientType,
+                        dataStoreName: this._dataStoreName,
+                        bodyType,
+                    });
+                    return cbOnce(errorInstances.InternalError
+                        .customizeDescription('Unsupported response body type returned from AWS SDK'));
+                }
+
+                if (data.$metadata?.httpHeaders) {
+                    log.trace(`${this.type} GET request response headers`, {
+                        responseHeaders: data.$metadata.httpHeaders,
+                        backendType: this.clientType,
+                    });
+                }
+
+                let finished = false;
+                const originalDestroy = stream.destroy.bind(stream);
+
+                const abortRequest = err => {
+                    if (finished) {
+                        return stream;
+                    }
+                    finished = true;
+                    log.debug('aborting GET request in progress', { objectGetInfo });
+                    abortController.abort(err);
+                    if (rawBody && rawBody !== stream) {
+                        rawBody.destroy?.(err);
+                    }
+                    try {
+                        originalDestroy(err);
+                    } catch (destroyErr) {
+                        log.debug('error while destroying aws sdk stream', {
+                            method: 'AwsClient.get',
+                            backendType: this.clientType,
+                            dataStoreName: this._dataStoreName,
+                            error: destroyErr,
+                        });
+                    }
+                    return stream;
+                };
+
+                stream.abort = abortRequest;
+                stream.destroy = err => abortRequest(err);
+                stream.createReadStream = () => stream;
+
```

Comment:
somehow here we are recreating the whole stream : instead of modifying the "normalized" object returned by AWS SDK, should we not just create the appropriate "stream" (as we need for our use-case) which will return AWS' stream in its `createReadStream` method?

```suggestion
const stream = {
    abort: abortRequest,
    destroy: abortRequest,
    createReadStream: () => data.Body,
};
```